### ISSUE TITLE
margins cut for PostScript printing (Okular/KDE)

### ISSUE THREAD

**Issue by kolAflash** [OpenPrinting/cups-filters#250]
Margins are being cut for many documents when printing in PostScript mode via Okular.
Tried a lot of settings in Okular (KDE Desktop document viewer which sends PostScript to CUPS), but that made no difference.

Evince gets everything correct, but Okular developers says it's due to Evince (Gnome document viewing tool) sending PDF to CUPS instead of PostScript.
https://bugs.kde.org/show_bug.cgi?id=422690#c5

Currently I can only reproduce this on my HP DeskJet ink printer at home.
But I also had this issue month before on a Canon laser printer (currently I'm not in office because of Covid-19, so I can't access that Canon).

I've tested this with Debian and openSUSE. And PostScript printing with Okular always cuts margins and PDF printing with Evince always keeps margins.

Re-printing Okular jobs with `lp` from command line doesn't change the result. See for details:
https://bugs.kde.org/show_bug.cgi?id=422690#c4
 
Only enabling "Force rasterization" (Okular printing settings) workarounded the issue for Okular.
&nbsp;

Please see attachments for details:
- scanned printouts with cut margins
- /var/spool/cups/
- cupsd.conf
[okular_cuts_margins_20200609.zip](https://github.com/OpenPrinting/cups-filters/files/4813168/okular_cuts_margins_20200609.zip)
- printed on German ISO/DIN A4 standard paper.
- Debian-Testing (11 a.k.a Bulleye 2020-06-03)
  - Okular-20.04 (a.k.a 1.10.0)
  - Evince-3.36.1
  - Printer: HP Deskjet 1000 j110 Series, hpcups 3.20.5
- openSUSE-15.1
  - Okular 18.12.3 (1.6.3)
  - Evince 3.26.0
  - Printer: HP Deskjet 1000 j110 Series, hpcups 3.12.11

localization: 
Both os (openSUSE and Debian) run on German localization.
But for testing on Debian I started okular with LANG=C.UTF-8.
&nbsp;

**cut margins (Okular via PostScript)**
![cut_margins_okular](https://user-images.githubusercontent.com/3355089/85290816-e7ca1d00-b499-11ea-9264-482b9dff9886.jpeg)

**with margins (Evince via PDF)**
![with_margins_evince](https://user-images.githubusercontent.com/3355089/85290908-0cbe9000-b49a-11ea-863b-4221d25f2229.jpeg)

**Comment by michaelweghorn** [comment]
> Re-printing Okular jobs with lp from command line doesn't change the result. See for details:
https://bugs.kde.org/show_bug.cgi?id=422690#c4

In other words, the issue can be reproduced without Okular by just printing the (Okular-generated) PostScript file using command line:

    lp -d HP_DeskJet_1000_at_fbox_hpcups -o fit-to-page debian_11-beta-20200609/var/spool/cups/d00018-001_-_print-area

The PostScript file itself looks just fine when e.g. opened in Okular.

**Comment by trialuser02** [comment]
[gstoraster-disable-hwmargins-for-pdf.patch.txt](https://github.com/OpenPrinting/cups-filters/files/4930662/gstoraster-disable-hwmargins-for-pdf.patch.txt)

I have tried to disable HWMargins for PDF output. See attached patch. Unfortunately, I don't know is it correct, but it works for me.

**Comment by tillkamppeter** [comment]
Fixed in commit e4f0f09e2a.

@trialuser02, thank you very much for your patch. I have applied it now. This patch skips margin handling when output is PDF and this is what we actually need, as with PDF as output the filter is used to turn incoming PostScript into PDF, before applying `pdftopdf`, so margin handling is handled later and should not get duplicated here. With PCL-XL or CUPS Raster as output format this filter converts PDF from the `pdftopdf` filter to prepare the pages for the printer, this is the step where the margins get applied.

Also thanks to @kolAflash for reporting this bug.
